### ISSUE TITLE
Time series: stacking bug

### ISSUE THREAD

**Issue by leeoniya** [grafana/grafana#34912]
reported by @wardbekker 

https://wardbekker.grafana.net/d/UlplAQqMk/nginx-overview-update?orgId=1&from=1622104554603&to=1622110271447&viewPanel=2

![image](https://user-images.githubusercontent.com/43234/120004910-f029b280-bf9c-11eb-86ed-2c25ad61d938.png)

![image](https://user-images.githubusercontent.com/43234/120004793-ce303000-bf9c-11eb-9ea9-8c2b6b1620ea.png)

**Comment by ploquets** [comment]
The stack in time series was working fine in 7.4.0 
After this version, we start to see weird stack, like this:
![image](https://user-images.githubusercontent.com/5149178/124682985-7d1c3180-dea2-11eb-92e9-3a5b509d4a0f.png)

**Comment by ploquets** [comment]
Tried on 8.0.4 and this behavior is still there.

**Comment by leeoniya** [comment]
that looks very odd (and different than what's in this bug report), but you've given us no information to investigate it for you.

what's your OS, browser, dashboard config, data response, etc.?

does this look correct to you?

https://play.grafana.org/d/000000012/grafana-play-home?orgId=1&editPanel=2

**Comment by ploquets** [comment]
This is stack on 8.0.4:

Stack OFF:
![image](https://user-images.githubusercontent.com/5149178/124683481-8063ed00-dea3-11eb-84e3-56e020a8c20d.png)

When Stack is set to normal:
![image](https://user-images.githubusercontent.com/5149178/124683522-94a7ea00-dea3-11eb-88f3-b34ab0fb54de.png)

Was working OK (stacking values normally) on 7.4.0

**Comment by leeoniya** [comment]
looks like same situation as this: https://github.com/grafana/grafana/issues/35554#issuecomment-873671806

**Comment by ploquets** [comment]
I don´t know if this is really true, because on 7.4.0 seems to work ok.

**Comment by sylock** [comment]
Any news from the devs? It makes some graphs unusable anymore after the update. It is a regression. Can you bring some attention to it?

**Comment by sylock** [comment]
After some time to try to reproduce on play.grafana.org here the steps to reproduce the issue:

## Explanation about the original data
- grafana 8.4.5
- zabbix datasource 4.2.5
- I configured multiple queries to get different CPU metrics (idle time, user, nice, system ...)
- I got the issue reported in this ticket as soon as I configured the graph as being a stacked one
- Then to try to reproduce the issue I exported the data into CSV, imported them into google sheet and share them without authentication

## Reproduce
- go on play.grafana.org
- create a new time series graph
- create 2 datasources google sheets (with my data stored into google sheets)
  - CPU nice time sheet id: 1su9i6yhVmhqtWG8B6MwPqBIucBGsjbFRk_iq3uqdzIA
  - CPU idle time sheet id: 1urFlTY2qz4yAQ2X5X98DKI_rjUFjRC05RJKTfbm9a8Q
- you apply 2 transformations to convert the data type of the value from string to numeric
- you select this timestamp range (around the data source available metrics)
  - UTC
  - from: 2022-04-05 17:40:00
  - to: 2022-04-05 18:41:00
- now you have a regular time series graph
- now you choose "normal" for the "stack" option
- you'll the same issue than reported in this ticket : the stack is not working

Then I tried to reproduce on a Grafana version 7.5.15
- I could not use the google sheet (the plugin is older and I could not configure it to work
- I configured the zabbix data source (version 4.1.5)
- I can see that everything is working fine

When I try to see what's the difference between
- config 1 : grafana 8.4.5 + zabbix source 4.2.5
- config 2: grafana 7.2.15 + zabbix source 4.1.5
I can see in the data table that : 
- option 1 : you get not aligned timestamps from the different series
- option 2: you get aligned timestemps from the different series

So the issue could be that the recent zabbix data source fails to align metrics coming from different queries related to the same graph.

**Comment by dprokop** [comment]
Indeed, this issue is releated to time-field missaligned series. It can be easily reprodued using a very basic dataset:

Series A:
```
"Time","CPU 1"
1649148066000,1
1649148126000,2
```

Series B:
```
"Time","CPU idle"
1649148123000,90
1649148183000,100
```

The resulting viz is this:

<img width="1611" alt="image" src="https://user-images.githubusercontent.com/2376619/169271853-7983c2ca-5614-4808-9892-160cbf9ec147.png">

Currently, we are using 0 for the missing(undefined) value in such a scenario, which I think could be replaced by just undefined values - which I think should result in lines being connected correctly. @leeoniya you have spent plenty of time on the stacking in the recent weeks, do you think that would make sense?

**Comment by leeoniya** [comment]
you can try it, but i dont think this will do anything more than turn the behavior into a less obvious misbehavior. imo people expect the interpolated regions to be stacked, but this will still not be done even if we prevent the upper series from dropping down in mis-aligned points -- they will simply be connected to the next mis-aligned point in the same series, effectively rendering as unstacked.

one option here might be to create a new aggregating transform that will bucket & sum at configurable time increments, e.g. 15s. this will reduce the datapoint count and make it less sensitive to mis-alignments within the chosen bucket size. this wouldn't work for scenarios where the metric is not something sum-able like "frequency", though we can make this transform do other aggregations such as avg, or geometric mean, high, low, etc.

**Comment by codeincarnate** [comment]
Reading through past and linked comments it looks like bucketing at one point happened as a part of null value insertion although that was a side effect and not intended to do time bucketing to fix stacking issues. Seems like this is an edge case and we'll revisit as time allows.

**Comment by UserNotFound** [comment]
I am noticing the same behavior.

Normal series data:
![image](https://user-images.githubusercontent.com/1580788/205988957-b2c8bec0-776a-476c-a17f-33553772c832.png)

Any attempt to use stacking seems to assume all null values are zero, no matter what I set for "Connect null values"?
![image](https://user-images.githubusercontent.com/1580788/205989055-e74431ea-f3e2-49b6-9391-6bc9c9f0d01d.png)

Detailed zoom:
![image](https://user-images.githubusercontent.com/1580788/205989487-18ae855c-b9aa-4b20-97f8-61d8da1618eb.png)

All I want is to see the total without rewriting the query, which stacking does in other tools like Kibana and New Relic, even when the source series data needs to be interpolated.

**Comment by leeoniya** [comment]
> which stacking does in other tools like Kibana and New Relic, even when the source series data needs to be interpolated.

out of curiosity, what does the tooltip in these tools say when you hover the interpolated regions? do they show up as if there is actual data there with the interpolated value? i'm wondering if they interpolate during rendering or during data pre-processing.

is there some public Kibana or New Relic stacked area panel i can play around with?

**Comment by BrainBug0815** [comment]
> out of curiosity, what does the tooltip in these tools say when you hover the interpolated regions?
You get either data for Series A or Series B. No interpolated values.

**Comment by BrainBug0815** [comment]
found a workaround:
group by time (e.g. "10s") and use an aggregation function (mean, sum,...)
=> this makes data points every 10s which are then aligned and stackable.

this does not work with Selectors (e.g. "top") as in that case, the timestamp of the top value is used and not "rounded".

**Comment by danistst** [comment]
tried this workaround, but I still get the "non stacking" behaviour.

**Comment by gitneep** [comment]
stacking works for me, but changing the order via the legend doesn't change the order in which the stack is rendered, which was working in the old time series graph, which is now no longer available (9.4.7)

https://github.com/grafana/grafana/issues/30907#issuecomment-1484656875

**Comment by retrodaredevil** [comment]
Anyone make any progress on this issue? I had been on an old Grafana version for a while (7.0.3) and it was working fine then, but after making the big jump to Grafana 10.3.3, I'm experiencing the issues described here.

**Comment by retrodaredevil** [comment]
Hello GitHub actions bot. I am writing to inform you that I believe this issue to still be relevant. 

I hope that you agree. 

Kind regards, 
Lavender
